### PR TITLE
Set provisioning API error message more informative (#56387)

### LINKED ISSUE #56387: Provisioning API returns generic error for invalid duration
**What happened**:

The Provisioning API returns the following generic error when the `For` field is invalid:

```
{"message":"bad request data","traceID":""}
```

The error is logged here:

```
ERROR[10-05|11:21:33] bad request data logger=context userId=1 orgId=1 uname=admin error="not a valid duration string: \"300\"" remote_addr=127.0.0.1 traceID=
```

However it should also be sent in the response.

**What you expected to happen**:

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- Grafana version:
- Data source type & version:
- OS Grafana is installed on:
- User OS & Browser:
- Grafana plugins:
- Others:

### PR DESCRIPTION
**Pull request by davidifr**
**What is this feature?**

Set a more informative error message string to the provisioning API in production mode.

**Why do we need this feature?**

Users in production mode needs to know what part of their request is causing the bad request error.
Just sending a message saying "bad request data" is not enough.

**Who is this feature for?**

[Add information on what kind of user the feature is for.]

**Which issue(s) does this PR fix?**:

Fixes #56387

**Special notes for your reviewer**:

### CODE REVIEW

**Review by brendamuir — APPROVED** [review]

**Review by grobinson-grafana — CHANGES_REQUESTED** [review]
Hi! 👋 I'm afraid this pull request edits a generated file, I think you want to edit `pkg/services/ngalert/api/tooling/go/api_provisioning.go` and run `make gen`.

**Comment by davidifr** [comment]
Hi @grobinson-grafana,
Sorry for the ignorance but I couldn't find the go directory under the [tooling directory](https://github.com/grafana/grafana/tree/main/pkg/services/ngalert/api/tooling) .
Is it possible you meant pkg/services/ngalert/api/api_provisioning.go ?

**Comment by grobinson-grafana** [comment]
Apologies, I think it might be `services/ngalert/api/tooling/swagger-codegen/templates/controller-api.mustache`:

```
if err := web.Bind(ctx.Req, &conf); err != nil {
	return response.Error(http.StatusBadRequest, "bad request data", err)
}
```

**Comment by alexweav** [comment]
@davidifr That particular string is hardcoded in a template file, in `pkg/services/ngalert/api/tooling/swagger-codegen/templates/controller-api.mustache`, line 28. Then, `make` in the `tooling` dir. Also, note that this will change **every endpoint in alerting**, since all of them use that template.

File: pkg/services/ngalert/api/generated_base_api_provisioning.go [hunk 1]
```diff
@@ -100,23 +100,23 @@ func (f *ProvisioningApiHandler) RoutePostAlertRule(ctx *models.ReqContext) resp
 	// Parse Request Body
 	conf := apimodels.ProvisionedAlertRule{}
 	if err := web.Bind(ctx.Req, &conf); err != nil {
-		return response.Error(http.StatusBadRequest, "bad request data", err)
+		return response.Error(http.StatusBadRequest, err.Error(), err)
```
Comment (alexweav) [inline pkg/services/ngalert/api/generated_base_api_provisioning.go:103]: So the API already does return this error string, just in a separate field.

For example, when I force a deserialization error, I get something like:
```
{
  "error": "json: cannot unmarshal number into Go struct field EmbeddedContactPoint.type of type string",
  "message": "bad request data",
  "traceID": ""
}
```
The string we're looking for is already being returned, and adding it here would duplicate it. Did we see a place where the `error` string was not being correctly displayed to the user? If so, that feels like a bug in the presentation layer and not here.

Comment (davidifr) [reply pkg/services/ngalert/api/generated_base_api_provisioning.go:103]: According to #56387 the string is being returned when running in non-production mode due to this [condition ](https://github.com/grafana/grafana/blob/53baecd71f5519e0ef300f3067cb89eab2908e2f/pkg/api/response/response.go#L213) .
The fix here should provide a proper error string under the `message` field when running in production mode.

Comment (alexweav) [reply pkg/services/ngalert/api/generated_base_api_provisioning.go:103]: Wow, that's unfortunate that there is such a difference between debug and release builds. Nice find!

**Comment by grobinson-grafana** [comment]
Hi @davidifr! Would it be possible to fix the lint errors? You can find them in `continous-integration/drone/pr` and `lint-backend`.

**Pull request closed without merging**